File: avalonia/ClaudeUsage.Tests/PricingTests.cs
using ClaudeUsage.Models;
using ClaudeUsage.Services;
using Xunit;

namespace ClaudeUsage.Tests;

public class PricingTests
{
    [Theory]
    [InlineData("claude-opus-5", 5d, 25d)]
    [InlineData("claude-opus-4-8", 5d, 25d)]
    [InlineData("claude-opus-4-7", 5d, 25d)]
    [InlineData("claude-opus-4-6", 5d, 25d)]
    [InlineData("claude-opus-4-5", 5d, 25d)]
    [InlineData("claude-fable-5", 10d, 50d)]
    [InlineData("claude-mythos-5", 10d, 50d)]
    [InlineData("claude-haiku-4-5-20251001", 1d, 5d)]
    // Sonnet 5 is cheaper than Sonnet 4.6 — pricing it at the older family rate
    // overstates its cost by half.
    [InlineData("claude-sonnet-5", 2d, 10d)]
    [InlineData("claude-sonnet-4-6", 3d, 15d)]
    [InlineData("claude-sonnet-4-5-20250929", 3d, 15d)]
    // Opus billed $15/$75 before the 4.5 price drop.
    [InlineData("claude-3-opus-20240229", 15d, 75d)]
    [InlineData("claude-opus-4-20250514", 15d, 75d)]
    [InlineData("claude-opus-4-1-20250805", 15d, 75d)]
    public void KnownModelsHaveTheirPostedRates(string modelId, double input, double output)
    {
        Assert.True(Pricing.TryGetPricing(modelId, out var p), $"no rate found for {modelId}");
        Assert.Equal(input, p.Input);
        Assert.Equal(output, p.Output);
    }

    [Theory]
    [InlineData("gpt-4o")]
    [InlineData("llama-3.1-70b")]
    [InlineData("")]
    public void UnknownModelHasNoRate(string modelId)
    {
        // A model routed through a non-Anthropic provider must not be priced as
        // though it were Claude.
        Assert.False(Pricing.TryGetPricing(modelId, out _));
    }

    [Fact]
    public void CacheRatesAreMultiplesOfInput()
    {
        Assert.True(Pricing.TryGetPricing("claude-opus-5", out var p));
        Assert.Equal(6.25, p.CacheWrite5m);
        Assert.Equal(10.0, p.CacheWrite1h);
        Assert.Equal(0.5, p.CacheRead);
    }

    [Fact]
    public void CostSumsEveryTokenKindAtItsOwnRate()
    {
        // One million of each token type on Opus 5, so each term reads as its
        // own rate: 5 + 25 + 6.25 + 10 + 0.5.
        var usage = new ModelUsage
        {
            InputTokens = 1_000_000,
            OutputTokens = 1_000_000,
            CacheCreation5mTokens = 1_000_000,
            CacheCreation1hTokens = 1_000_000,
            CacheReadInputTokens = 1_000_000,
        };

        Assert.True(Pricing.TryGetCost("claude-opus-5", usage, out var cost));
        Assert.Equal(46.75, cost, 9);
    }

    [Fact]
    public void CacheWriteTtlsArePricedApart()
    {
        // The same cache-write volume costs more on the 1-hour TTL. Collapsing
        // the two into one 1.25x bucket, as a single cache_creation rate would,
        // understates the 1-hour case.
        Pricing.TryGetCost("claude-opus-5", new ModelUsage { CacheCreation5mTokens = 1_000_000 }, out var cost5m);
        Pricing.TryGetCost("claude-opus-5", new ModelUsage { CacheCreation1hTokens = 1_000_000 }, out var cost1h);

        Assert.Equal(6.25, cost5m);
        Assert.Equal(10.0, cost1h);
    }

    [Fact]
    public void UnknownModelCostsNothingAndIsFlagged()
    {
        var usage = new ModelUsage { InputTokens = 1_000_000, OutputTokens = 1_000_000 };

        Assert.False(Pricing.TryGetCost("some-other-provider/model", usage, out var cost));
        Assert.Equal(0d, cost);
    }
}

File: swiftui/ClaudeUsage/Tests/ClaudeUsageTests/PricingTests.swift
import Foundation
import Testing

@testable import ClaudeUsage

@Suite("Pricing")
struct PricingTests {

    @Test("Known models bill at their posted rates", arguments: [
        ("claude-opus-5", 5.0, 25.0),
        ("claude-opus-4-8", 5.0, 25.0),
        ("claude-opus-4-7", 5.0, 25.0),
        ("claude-opus-4-6", 5.0, 25.0),
        ("claude-opus-4-5", 5.0, 25.0),
        ("claude-fable-5", 10.0, 50.0),
        ("claude-mythos-5", 10.0, 50.0),
        ("claude-haiku-4-5-20251001", 1.0, 5.0),

        // Sonnet 5 is cheaper than Sonnet 4.6 — pricing it at the older family
        // rate overstates its cost by half.
        ("claude-sonnet-5", 2.0, 10.0),
        ("claude-sonnet-4-6", 3.0, 15.0),
        ("claude-sonnet-4-5-20250929", 3.0, 15.0),

        // Opus billed $15/$75 before the 4.5 price drop.
        ("claude-3-opus-20240229", 15.0, 75.0),
        ("claude-opus-4-20250514", 15.0, 75.0),
        ("claude-opus-4-1-20250805", 15.0, 75.0),
    ])
    func ratesForKnownModels(modelID: String, input: Double, output: Double) throws {
        let rates = try #require(Pricing.rates(for: modelID), "no rate found for \(modelID)")
        #expect(rates == ModelPricing(input: input, output: output))
    }

    @Test("A model routed elsewhere has no Claude rate", arguments: ["gpt-4o", "llama-3.1-70b", ""])
    func ratesForUnknownModel(modelID: String) {
        #expect(Pricing.rates(for: modelID) == nil)
    }

    @Test("Cache traffic bills as a multiple of the input rate")
    func cacheMultipliers() throws {
        let rates = try #require(Pricing.rates(for: "claude-opus-5"))
        #expect(rates.cacheWrite5m == 6.25)
        #expect(rates.cacheWrite1h == 10)
        #expect(rates.cacheRead == 0.5)
    }

    @Test("Each token type is priced at its own rate")
    func cost() throws {
        // One million of each token type on Opus 5, so each term reads as its
        // own rate: 5 + 25 + 6.25 + 10 + 0.5.
        let usage = ModelUsage(
            inputTokens: 1_000_000,
            outputTokens: 1_000_000,
            cacheReadInputTokens: 1_000_000,
            cacheCreation5mTokens: 1_000_000,
            cacheCreation1hTokens: 1_000_000
        )

        let cost = try #require(Pricing.cost(for: "claude-opus-5", usage: usage))
        #expect(abs(cost - 46.75) < 1e-9)
    }

    @Test("The two cache TTLs bill differently")
    func costSeparatesCacheTTLs() throws {
        // Collapsing the two into one 1.25x bucket, as a single cache_creation
        // rate would, understates the 1-hour case.
        let write5m = ModelUsage(cacheCreation5mTokens: 1_000_000)
        let write1h = ModelUsage(cacheCreation1hTokens: 1_000_000)

        #expect(Pricing.cost(for: "claude-opus-5", usage: write5m) == 6.25)
        #expect(Pricing.cost(for: "claude-opus-5", usage: write1h) == 10)
    }

    @Test("An unpriced model has no cost rather than a zero one")
    func costOfUnknownModelIsUnknown() {
        let usage = ModelUsage(inputTokens: 1_000_000, outputTokens: 1_000_000)
        #expect(Pricing.cost(for: "some-other-provider/model", usage: usage) == nil)
    }

    // Cost rendering lives in the "Cost formatting" suite.

    @Test("A specific model version beats its family in the name table")
    func friendlyModelNames() {
        #expect(Formatting.friendlyModel("claude-sonnet-5") == "Sonnet 5")
        #expect(Formatting.friendlyModel("claude-opus-4-8") == "Opus 4.8")
        #expect(Formatting.friendlyModel("claude-opus-4-20250514") == "Opus 4")
        #expect(Formatting.friendlyModel("claude-fable-5") == "Fable 5")
        #expect(Formatting.friendlyModel("gpt-4o") == "gpt-4o")
    }
}
